File: python/euler/bin/p0021.py
# project euler: problem 21

from euler.lib.util import aliquot_sum_tbl


def compute(num: int) -> str:
    tbl = aliquot_sum_tbl(num - 1)

    return str(
        sum(x + tbl[x] for x in range(2, len(tbl)) if x > tbl[x] and tbl[tbl[x]] == x)
    )


def solve() -> str:
    return compute(10_000)

File: typescript/euler/bin/p0021.ts
// project euler: problem 21

import { aliquotSumTbl } from "../lib/factor.ts";
import { sum } from "../lib/math.ts";
import { range } from "../lib/util.ts";

export const compute = (n: number): string => {
  const tbl = aliquotSumTbl(n - 1);
  const n_set = new Set();
  for (const i of range(2, n)) {
    if (n_set.has(i) === true) {
      continue;
    }
    const d1 = tbl[i];
    if (d1 <= i) {
      continue;
    }
    const d2 = tbl[d1];
    if (i === d2) {
      n_set.add(d1);
      n_set.add(d2);
    }
  }

  return String(sum(Array.from(n_set) as number[]));
};

export const solve = (): string => compute(10_000);
